engine-rest/engine-rest/src/main/java/org/camunda/bpm/engine/rest/impl/ProcessInstanceRestServiceImpl.java

```diff
@@ -84,8 +85,18 @@ public List<ProcessInstanceDto> queryProcessInstances(
 
     List<ProcessInstanceDto> instanceResults = new ArrayList<>();
     for (ProcessInstance instance : matchingInstances) {
-      ProcessInstanceDto resultInstance = ProcessInstanceDto.fromProcessInstance(instance);
-      instanceResults.add(resultInstance);
+      if (null!= queryDto.isWithVariablesInReturn() && queryDto.isWithVariablesInReturn()){
+        RuntimeService runtimeService = engine.getRuntimeService();       
+        VariableMap variableMap = (VariableMap) runtimeService.getVariables(instance.getProcessInstanceId());
```

Comment:
Querying all the variable instances within a single query would improve the performance.
Can you give it a try? 
I think you can get the VariableInstances similar to this:
```
    List<VariableInstance> variableInstances =
        runtimeService.createVariableInstanceQuery().activityInstanceIdIn(
            matchingInstances.stream().map(ProcessInstance::getId).toArray(String[]::new)
        ).list();
```
then you need to match the VariableInstances to the Instances